File: ios/Classes/APlayer.swift
//
//  APlayer.swift
//  a_player
//
//  Created by drm-chao wang on 2022/6/23.
//

import Foundation
import Flutter

class APlayer: NSObject, FlutterTexture, FlutterStreamHandler, APlayerListener {
    
    private let queuingEventSink: QueuingEventSink = QueuingEventSink.init()
  
    private var eventChannel: FlutterEventChannel?
    private var methodChannel: FlutterMethodChannel?
    private var latestBuffer: CVImageBuffer!
    private var registrar: FlutterPluginRegistrar!
    private var _textureId: Int64!
    private var textureRegistry: FlutterTextureRegistry?
    private var player: APlayerInterface?
    private var kernel: Int?
    
    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        self.textureRegistry = registrar.textures()
        super.init()
        self.bindFlutter()
    }
    
    
    func textureId() -> Int64 {
        return _textureId
    }
    private func bindFlutter() -> Void {
        
        _textureId = textureRegistry?.register(self)
        
        eventChannel = FlutterEventChannel.init(name: PLAYER_EVENT_CHANNEL_NAME + String(_textureId), binaryMessenger: registrar.messenger())
        methodChannel = FlutterMethodChannel.init(name: PLAYER_METHOD_CHANNEL_NAME + String(_textureId), binaryMessenger: registrar.messenger())
        
        self.eventChannel?.setStreamHandler(self)
        self.methodChannel?.setMethodCallHandler {[weak self] (call: FlutterMethodCall, result: FlutterResult) -> Void in
           
            switch call.method {
              case "play":
                self?.play()
                break
              case "pause":
                self?.pause()
                break
              case "stop":
                self?.stop()
                break
              case "prepare":
                self?.prepare()
                break
              case "setDataSource":
                self?.setDataSource(args: call.arguments as! Dictionary<String, Any>)
                break
            case "restart":
                self?.restartPlay()
                break
              case "seekTo":
                self?.seekTo(position: call.arguments as! Int64)
                break
              case "setSpeed":
                self?.setSpeed(speed: call.arguments as! Float)
                break
              case "setLoop":
                self?.setLoop(loop: call.arguments as! Bool)
                break
              case "release":
                self?.destroy()
                break
              default:
                break
            }
            result(nil)
        }
    }
    private func setupPlayer() -> Void {
        player?.addListener(listener: self)
    }
    
    private func createPlayer() -> Void {
        queuingEventSink.success(event: [
            "type": "initializing"
        ])
        resetValue()
        player?.destroy()
        player = nil
        switch(kernel) {
        case KERNEL_ALIYUN:
            player = AliyunPlayerImpl.init()
            break
        case KERNEL_IJK:
            player = IJKPlayerImpl.init()
            break
        case KERNEL_AV:
            player = AVPlayerImpl.init()
            break
        default:
            break
        }
        if (player == nil) {
            return
        }
        setupPlayer()
        queuingEventSink.success(event: [
            "type": "initialized",
            "data": self.kernel
        ])
    }
    
    private func resetValue() -> Void {
        latestBuffer = nil
        stop();
    }
    private func restartPlay() -> Void {
        seekTo(position: 0)
        play()
    }
    private func setDataSource(args: Dictionary<String, Any>) -> Void {
        resetValue()
        let kernel = args["kernel"] as! Int
        let url = args["url"] as! String
        let position = args["position"] as! Int64
        if (APlayerUtil.isHttpProtocol(url: url) == true) {
            player?.setHttpDataSource(url: url, startAtPositionMs: position, headers: args["httpHeaders"] as! Dictionary<String, String>)
        } else if (APlayerUtil.isFileProtocol(url: url) == true) {
            player?.setFileDataSource(path: url, startAtPositionMs: position)
        }
        self.setKernel(kernel: kernel)
        
    }
    private func setKernel(kernel: Int) {
        if (self.kernel == kernel) {
            return
        }
        self.kernel = kernel
        createPlayer()
    }
    private func prepare() -> Void {
        player?.prepare()
    }
    
    private func play() -> Void {
        player?.play()
    }
    
    private func pause() -> Void {
        player?.pause()
    }
    private func stop() -> Void {
        player?.stop()
    }
    private func seekTo(position: Int64) -> Void {
        player?.seekTo(positionMs: position)
    }
    private func setSpeed(speed: Float) -> Void {
        player?.setSpeed(speed: speed)
    }
    
    private func setLoop(loop: Bool) -> Void {
        player?.setLoop(isLoop: loop)
    }
    private func destroy() -> Void {
        player?.destroy()
        player = nil
        queuingEventSink.endOfStream()
        textureRegistry = nil
        eventChannel = nil
        methodChannel = nil
    }
    
    func copyPixelBuffer() -> Unmanaged<CVPixelBuffer>? {
        if latestBuffer == nil {
            return nil
        }
        return Unmanaged<CVPixelBuffer>.passRetained(latestBuffer)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        queuingEventSink.setDelegate(delegate: events)
        createPlayer()
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        queuingEventSink.setDelegate(delegate: nil)
        return nil
    }
    func onPixelBuffer(pixelBuffer: CVPixelBuffer) {
        if (latestBuffer == pixelBuffer) {
            return
        }
        latestBuffer = pixelBuffer
        textureRegistry?.textureFrameAvailable(_textureId)
    }
    
    func onPlayingListener(isPlaying: Bool) {
        queuingEventSink.success(event: [
            "type": "playing",
            "data": isPlaying
        ])
    }
    
    func onErrorListener(code: String, message: String) {
        queuingEventSink.success(event: [
            "type": "error",
            "data": "\(code): \(message)"
        ])
    }
    
    func onCompletionListener() {
        queuingEventSink.success(event: [
            "type": "completion",
        ])
    }
    
    func onReadyToPlayListener() {
        queuingEventSink.success(event: [
            "type": "readyToPlay",
            "data": [
                "duration": player!.duration,
                "playSpeed": player!.speed
            ]
        ])
    }
    
    func onVideoSizeChangedListener(width: Int32, height: Int32) {
        queuingEventSink.success(event: [
            "type": "videoSizeChanged",
            "data": [
                "height": height,
                "width": width
            ]
        ])
    }
    
    func onCurrentPositionChangedListener(position: Int64) {
        queuingEventSink.success(event: [
            "type": "currentPositionChanged",
            "data": position
        ])
    }
    
    func onCurrentDownloadSpeedChangedListener(speed: Int64) {
        queuingEventSink.success(event: [
            "type": "currentDownloadSpeedChanged",
            "data": speed
        ])
    }
    
    func onBufferedPositionChangedListener(buffered: Int64) {
        queuingEventSink.success(event: [
            "type": "bufferedPositionChanged",
            "data": buffered
        ])
    }
    
    func onLoadingBeginListener() {
        queuingEventSink.success(event: [
            "type": "loadingBegin"
        ])
    }
    
    func onLoadingProgressListener(percent: Int) {
        queuingEventSink.success(event: [
            "type": "loadingProgress",
            "data": percent
        ])
    }
    
    func onLoadingEndListener() {
        queuingEventSink.success(event: [
            "type": "loadingEnd"
        ])
    }
}

File: android/src/main/kotlin/tech/shmy/a_player/APlayer.kt
package tech.shmy.a_player

import android.app.Activity
import android.app.PictureInPictureParams
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.Rect
import android.graphics.SurfaceTexture
import android.os.Build
import android.provider.Settings
import android.util.Rational
import android.view.Surface
import android.widget.Toast
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import tech.shmy.a_player.player.*
import tech.shmy.a_player.player.impl.AliyunPlayerImpl
import tech.shmy.a_player.player.impl.ExoPlayerImpl
import tech.shmy.a_player.player.impl.IJKPlayerImpl


class APlayer(
    private val context: Context,
    private val activity: Activity,
    private val textureEntry: TextureRegistry.SurfaceTextureEntry,
    private val binaryMessenger: BinaryMessenger,
) : EventChannel.StreamHandler {
    private var kernel: Int? = null;
    private var player: APlayerInterface? = null
    private val surfaceTexture: SurfaceTexture = textureEntry.surfaceTexture()
    private var surface: Surface? = null
    private val queuingEventSink: QueuingEventSink = QueuingEventSink()
    private var eventChannel: EventChannel? = null
    private var methodChannel: MethodChannel? = null

    init {
        bindFlutter()
    }

    private fun bindFlutter() {
        val textureId: Long = textureEntry.id()
        val eventChannel = EventChannel(binaryMessenger, "$PLAYER_EVENT_CHANNEL_NAME$textureId")
        val methodChannel = MethodChannel(binaryMessenger, "$PLAYER_METHOD_CHANNEL_NAME$textureId")
        eventChannel.setStreamHandler(this)
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "play" -> {
                    play()
                    result.success(null)
                }
                "pause" -> {
                    pause()
                    result.success(null)
                }
                "stop" -> {
                    stop()
                    result.success(null)
                }
                "prepare" -> {
                    prepare()
                    result.success(null)
                }
                "restart" -> {
                    restart()
                    result.success(null)
                }
                "setDataSource" -> {
                    setDataSource(call.arguments as MutableMap<String, Any>)
                    result.success(null)
                }
                "seekTo" -> {
                    seekTo((call.arguments as Int).toLong())
                    result.success(null)
                }
                "setSpeed" -> {
                    setSpeed((call.arguments as Double).toFloat())
                    result.success(null)
                }
                "setLoop" -> {
                    setLoop(call.arguments as Boolean)
                    result.success(null)
                }
                "enterPip" -> {
                    result.success(enterPip())
                }
                "openSettings" -> {
                    openSettings()
                    result.success(null)
                }
                "release" -> {
                    release()
                    result.success(null)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }

    }

    private fun createPlayer() {
        queuingEventSink.success(
            mapOf(
                "type" to "initializing"
            )
        )
        player?.release()
        player = null
        when (kernel) {
            KERNEL_ALIYUN -> {
                player = AliyunPlayerImpl.createPlayer(context)
            }
            KERNEL_IJK -> {
                player = IJKPlayerImpl.createPlayer(context)
            }
            KERNEL_EXO -> {
                player = ExoPlayerImpl.createPlayer(context)
            }
        }
        if (player == null) {
            return
        }
        setupPlayer()

        queuingEventSink.success(
            mapOf(
                "type" to "initialized",
                "data" to kernel
            )
        )
    }

    private fun setupPlayer() {
        player?.addListener(object : APlayerListener {
            override fun onVideoSizeChangedListener(width: Int, height: Int) {
                surfaceTexture.setDefaultBufferSize(width, height)
                queuingEventSink.success(
                    mapOf(
                        "type" to "videoSizeChanged",
                        "data" to mapOf(
                            "height" to height,
                            "width" to width,
                        )
                    )
                )
            }

            override fun onPlayingListener(isPlaying: Boolean) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "playing",
                        "data" to isPlaying,
                    )
                )
            }

            override fun onReadyToPlayListener() {
                queuingEventSink.success(
                    mapOf(
                        "type" to "readyToPlay",
                        "data" to mapOf(
                            "duration" to player!!.duration,
                            "playSpeed" to player!!.speed,
                        )
                    )
                )
            }

            override fun onErrorListener(code: String, message: String) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "error",
                        "data" to "$code: $message"
                    )
                )
            }

            override fun onCompletionListener() {
                queuingEventSink.success(
                    mapOf(
                        "type" to "completion",
                    )
                )
            }

            override fun onCurrentPositionChangedListener(position: Long) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "currentPositionChanged",
                        "data" to position
                    )
                )
            }

            override fun onCurrentDownloadSpeedChangedListener(speed: Long) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "currentDownloadSpeedChanged",
                        "data" to speed
                    )
                )
            }

            override fun onBufferedPositionChangedListener(buffered: Long) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "bufferedPositionChanged",
                        "data" to buffered
                    )
                )
            }

            override fun onLoadingBeginListener() {
                queuingEventSink.success(
                    mapOf(
                        "type" to "loadingBegin",
                    )
                )
            }

            override fun onLoadingProgressListener(percent: Int) {
                queuingEventSink.success(
                    mapOf(
                        "type" to "loadingProgress",
                        "data" to percent,
                    )
                )
            }

            override fun onLoadingEndListener() {
                queuingEventSink.success(
                    mapOf(
                        "type" to "loadingEnd",
                    )
                )
            }
        })
        setSurface()
    }

    private fun setSurface() {
        val canvas = surface?.lockCanvas(null)
        canvas?.drawColor(Color.BLACK)
        surface?.unlockCanvasAndPost(canvas)
        surface?.release()
        surface = Surface(surfaceTexture)
        player?.setSurface(surface!!)
    }

    private fun restart() {
        seekTo(0)
        play()
    }

    private fun setDataSource(
       args: Map<String, Any>
    ) {
        val kernel = args["kernel"] as Int
        val url = args["url"] as String
        val position = (args["position"] as Int).toLong()
        val httpHeaders = args["httpHeaders"] as Map<String, String>
        setKernel(kernel)
        if (APlayerUtil.isHttpProtocol(url)) {
            player?.setHttpDataSource(
                url,
                position, httpHeaders
            )
        } else if (APlayerUtil.isFileProtocol(url)) {
            player?.setFileDataSource(url, position)
        }
    }

    private fun prepare() {
        player?.prepare()
    }

    private fun play() {
        player?.play()
    }

    private fun pause() {
        player?.pause()
    }

    private fun stop() {
        player?.stop()
    }

    private fun setKernel(kernel: Int) {
        if (kernel == this.kernel) {
            return
        }
        this.kernel = kernel
        createPlayer()
    }

    private fun seekTo(position: Long) {
        player?.seekTo(position)
    }

    private fun setSpeed(speed: Float) {
        player?.setSpeed(speed)
    }

    private fun setLoop(isLoop: Boolean) {
        player?.setLoop(isLoop)
    }

    private fun release() {
        player?.release()
        player = null
        queuingEventSink.endOfStream()
        eventChannel?.setStreamHandler(null)
        methodChannel?.setMethodCallHandler(null)
        eventChannel = null
        methodChannel = null
    }

    private fun enterPip(): Boolean {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> {
                var builder: PictureInPictureParams.Builder =
                    PictureInPictureParams.Builder()
                        .setAspectRatio(Rational(16, 9))
                        .setSourceRectHint(Rect(0, 0, 0, 0))
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    builder = builder.setSeamlessResizeEnabled(true)
                }
                return activity.enterPictureInPictureMode(
                    builder.build()
                )
            }
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.N -> {
                activity.enterPictureInPictureMode()
                return true
            }
            else -> {
                return false
            }
        }
    }

    private fun openSettings() {
        Toast.makeText(
            context, "请在设置-画中画, 选择本App, 允许进入画中画模式", Toast.LENGTH_SHORT
        ).show()
        activity.startActivity(Intent(Settings.ACTION_SETTINGS))
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        queuingEventSink.setDelegate(events)
    }

    override fun onCancel(arguments: Any?) {
        queuingEventSink.setDelegate(null)
    }
}